src/api/firebase.js

```diff
@@ -92,16 +93,77 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
 export async function updateItem(listId, itemId, checked) {
 	const itemRef = doc(db, listId, itemId);
 	const itemDoc = await getDoc(itemRef);
-	const { dateLastPurchased, totalPurchases } = itemDoc.data();
+	const { dateCreated, dateLastPurchased, dateNextPurchased, totalPurchases } =
+		itemDoc.data();
 
-	try {
-		return updateDoc(itemRef, {
-			dateLastPurchased: checked ? new Date() : dateLastPurchased,
-			totalPurchases: checked ? totalPurchases + 1 : totalPurchases,
-			checked: checked,
-		});
-	} catch (error) {
-		console.log('updateDoc error', error);
+	const checkedDateLastPurchased = checked ? new Date() : dateLastPurchased;
+
+	if (checkedDateLastPurchased) {
+		// Convert dateNextPurchased from Timestamp to a JS Date
+		const dateNextPurchasedAsDate = dateNextPurchased.toDate();
+
+		/**
+		 * Calculate the previous estimate for purchase
+		 * based on historical data
+		 */
+		const previousEstimate = getDaysBetweenDates(
+			checkedDateLastPurchased,
+			dateNextPurchasedAsDate,
+		);
+
+		/**
+		 * Calculate the days since the last transaction,
+		 * considering either last purchased or the created date
+		 */
+		const daysSinceLastTransaction = getDaysBetweenDates(
+			new Date(),
+			checkedDateLastPurchased || dateCreated,
+		);
+
+		// Calculate the remaining days until the next purchase
+		let remainingDays = calculateEstimate(
+			previousEstimate,
+			daysSinceLastTransaction,
+			totalPurchases,
+		);
+
+		// Ensure that the next purchase date is at least one day in the future
+		if (remainingDays <= 0) {
+			remainingDays = 1;
+		}
+
+		if (dateNextPurchasedAsDate < checkedDateLastPurchased) {
+			// Set checked to false so the item can be purchased again
+			checked = false;
+		}
+
+		try {
+			return updateDoc(itemRef, {
+				dateLastPurchased: checkedDateLastPurchased,
+				dateNextPurchased: getFutureDate(remainingDays),
+				totalPurchases: totalPurchases + 1,
+				checked: checked,
+			});
+		} catch (error) {
+			console.log('updateDoc error', error);
+		}
+	} else {
+		try {
+			/**
+			 * Prevent dateNextPurchased from being overridden
+			 * with the current date or dateCreated
+			 */
+			const newDateNextPurchased =
+				dateNextPurchased && dateNextPurchased.toDate();
+
```

Comment:
This second `try` blocks looks similar to the first. Is there anything we can adjust to DRY it up?

Reply:
@mentalcaries I tried deleting the entire `else block` to DRY it up, and after testing the code, it seems to be working without it. However, it's strange because previously, after 24 hours had passed according to the implementation Shangguan and I did, *without* the `else block`, the code was unchecking the item in the frontend but keeping the `checked` property as `true` in the database.